cmd/oras/discover.go

```diff
@@ -0,0 +1,207 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+
+	"oras.land/oras-go/v2/registry/remote"
+	"oras.land/oras/cmd/oras/internal/option"
+
+	"github.com/need-being/go-tree"
+	digest "github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	artifactspec "github.com/oras-project/artifacts-spec/specs-go/v1"
+	"github.com/spf13/cobra"
+)
+
+type discoverOptions struct {
+	option.Common
+	option.Remote
+
+	targetRef    string
+	artifactType string
+	outputType   string
+}
+
+func discoverCmd() *cobra.Command {
+	var opts discoverOptions
+	cmd := &cobra.Command{
+		Hidden: true,
+		Use:    "discover [options] <name:tag|name@digest>",
+		Short:  "Discover artifacts from remote registry",
+		Long: `Discover artifacts from remote registry
+
+Example - Discover all the artifacts linked with the specified reference:
+  oras discover localhost:5000/hello
+
+Example - Discover all the artifacts linked with the specified reference in a tree view:
+  oras discover localhost:5000/hello -o tree
+
+Example - Discover artifacts of type test-artifact test-artifact linked with the specified reference:
+  oras discover --artifact test-artifact localhost:5000/hello
+`,
+		Args: cobra.ExactArgs(1),
+		PreRun: func(cmd *cobra.Command, args []string) {
+			fmt.Println("Command discover is in preview and might have breaking changes coming.")
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			opts.targetRef = args[0]
+			return runDiscover(opts)
+		},
+	}
+
+	cmd.Flags().StringVarP(&opts.artifactType, "artifact-type", "", "", "artifact type")
+	cmd.Flags().StringVarP(&opts.outputType, "output", "o", "table", "format in which to display references (table, json, or tree). tree format will show all references including nested")
```

Comment:
"format in which to display references (table, json, or tree)"
Should we use "references" or "referrers" here? @FeynmanZhou @yizha1 can suggest.

Reply:
Yeah CLI userd won't understand `predecessor`.